### ISSUE TITLE
Async Context?

### ISSUE THREAD

**Issue by ngbinh** [eikek/yamusca#34]
Right now, it is possible to load the context lazily as mentioned https://github.com/eikek/yamusca#lazy-context-example . But often time, the context is loaded from a database or other external sources. Is there a way to define 'AsyncContext` that wrapped `Value` on an effect type? It may come under another module that depends on [cat effects](https://typelevel.org/cats-effect/)

**Comment by eikek** [comment]
Hello,
this is currently not possible. I usually load all data from external sources like dbs into a case class and then render the template inside the same effect type. Where do you think is it good to have each value with an effect?

**Comment by ngbinh** [comment]
The main use case is to allow a more efficient lazy load.

**Comment by eikek** [comment]
I see, so this would be useful if the context is only partially used with each template. One could have one context backing multiple templates?

I think this means that a context is not `String => (Context, Option[Value])` but rather `String => F[(Context, Option[Value])]` right? Using `Id` for `F` would yield the current situation. I *think* that should be possible, but haven't looked at it more thoroughly. It sounds nice, but I won't be able to work on this in the near future. I'm happy to look at/help with PRs for this.

**Comment by ngbinh** [comment]
Yeah, I may try to take a bit at this. Do you have any pointers that would help me start? (may not be this week though)

**Comment by eikek** [comment]
Not really, I'm sorry :-/. I would have to dig into it first, too – it's been a while :-). A very high overview would be: the `expand.scala` contains the code that creates the final outcome (applying data to a template). Then there is a `Find` class in `context.scala` which is a state monad that extracts a value and the new context from a given context (which is solely based on the `Context` trait's `find` method). The expansion works by using a `Find` and instead of returning something call a callback function with the result (that concatenates the string outcome of each token, this is only for performance; to avoid concatenating a lot of strings a `StringBuilder` is used). So I would maybe start changing the `Context` trait, because this is the base of everything else. Then compile errors could guide you. But really, I didn't have a closer look, so take it with some salt….

Never worry about any timeline or something. I hope this helps a little. Just come back if you have specific questions, I might be of more help then.
